sdk/rust/oak/src/invocations.rs

```diff
@@ -0,0 +1,156 @@
+pub struct Invocation<
+    R: prost::Message + crate::handle::HandleVisit + crate::io::Decodable + std::default::Default,
+    S: prost::Message + crate::handle::HandleVisit + crate::io::Encodable + std::clone::Clone,
+> {
+    pub request_receiver: crate::io::Receiver<R>,
+    pub response_sender: crate::io::Sender<S>,
+}
+
+// TODO(#389): Automatically generate this code.
+impl<
+        R: prost::Message + crate::handle::HandleVisit + crate::io::Decodable + std::default::Default,
+        S: prost::Message + crate::handle::HandleVisit + crate::io::Encodable + std::clone::Clone,
+    > crate::io::Encodable for Invocation<R, S>
+{
+    fn encode(&self) -> Result<crate::io::Message, crate::OakError> {
+        let bytes = vec![];
+        let handles = vec![
+            self.request_receiver.handle.handle,
+            self.response_sender.handle.handle,
+        ];
+        Ok(crate::io::Message { bytes, handles })
+    }
+}
+
+// TODO(#389): Automatically generate this code.
+impl<
+        R: prost::Message + crate::handle::HandleVisit + crate::io::Decodable + std::default::Default,
+        S: prost::Message + crate::handle::HandleVisit + crate::io::Encodable + std::clone::Clone,
+    > crate::io::Decodable for Invocation<R, S>
+{
+    fn decode(message: &crate::io::Message) -> Result<Self, crate::OakError> {
+        if !message.bytes.is_empty() {
+            panic!(
+                "incorrect number of bytes received: {} (expected: 0)",
+                message.bytes.len()
+            );
+        }
+        if message.handles.len() != 2 {
+            panic!(
+                "incorrect number of handles received: {} (expected: 2)",
+                message.handles.len()
+            );
+        }
+        Ok(Invocation {
+            request_receiver: crate::io::Receiver::new(crate::ReadHandle {
+                handle: message.handles[0],
+            }),
+            response_sender: crate::io::Sender::new(crate::WriteHandle {
+                handle: message.handles[1],
+            }),
+        })
+    }
+}
+
+impl<
+        R: prost::Message + crate::handle::HandleVisit + crate::io::Decodable + std::default::Default,
+        S: prost::Message + crate::handle::HandleVisit + crate::io::Encodable + std::clone::Clone,
+    > Invocation<R, S>
+{
+    pub fn receive(&self) -> Result<R, crate::OakError> {
+        self.request_receiver.receive()
+    }
+
+    pub fn send(&self, response: &S) -> Result<(), crate::OakError> {
+        self.response_sender.send(response)
+    }
+
+    pub fn close_channels(&self) {
+        if let Err(error) = self.request_receiver.close() {
+            warn!("Error when trying to close request_receiver: {}", error);
+        }
+        if let Err(error) = self.response_sender.close() {
+            warn!("Error when trying to close response_sender: {}", error);
+        }
+    }
+}
+
+/// The write half of a channel sending external requests as `Invocations` to an Oak Node that
```

Comment:
I think the write half of the channel is quite deeply buried in the object structure Perhaps something like `Wrapper for a sender that can be used to send invocations to an Oak node that will handle the requests.`

Reply:
Thanks. Done.